algorithms/HPdclassifier/R/hpdegbm.R

```diff
@@ -0,0 +1,290 @@
+# bag.fraction: the fraction of the training set observations randomly selected for the next tree in the expansion
+# verbose: If TRUE, gbm prints out progress and performance indicators
+# trace: If TRUE, print out the training time
+# completeModel: If TRUE, add training time in the model 
+
+### start of hpdegbm function
+{
+   # parameters checking
+   if(missing(X_train))
+	stop("'X_train' is a required argument")
+
+   if(missing(Y_train))
+	stop("'Y_train' is a required argument")
+
+   if(!is.dframe(X_train) && !is.data.frame(X_train) && !is.darray(X_train) && !is.matrix(X_train))
+       stop("'X_train' must be a dframe or data.frame or darray or matrix")
+
+   if(!is.dframe(Y_train) && !is.data.frame(Y_train) && !is.darray(Y_train) && !is.matrix(Y_train) && !is.vector(Y_train))
+       stop("'Y_train' must be a dframe or data.frame or darray or matrix or numeric vector")
+
+   if(missing(dl_GBM_model))
+	stop("'dl_GBM_model' is a required argument")
+
+   if(missing(dbest.iter))
+	stop("'dbest.ier' is a required argument")
+
+   if(missing(nExecutor))
+	stop("'nExecutor' is a required argument")
+
+    nExecutor <- round(nExecutor)
+    if(nExecutor <= 0)
+        stop("nExecutor should be a positive integer number")
+
+   if(missing(distribution))
+	stop("'distribution' is a required argument")
+
+   if(n.trees <= 0)
+        stop("'n.trees' must be more than 0")
+
+   if(interaction.depth <= 0)
+        stop("'interaction.depth' must be more than 0")
+
+
+   if ((interaction.depth%%1 < 0) && ((interaction.depth%%1 > 0)))
+        stop("'interaction.depth' must be an integer")
+
+   if(n.minobsinnode <= 0)
+        stop("'n.minobsinnode' must be more than 0")
+
+   if ((n.minobsinnode%%1 < 0) && ((n.minobsinnode%%1 > 0)))
+        stop("'n.minobsinnode' must be an integer")
+
+   if(shrinkage <= 0)
+        stop("'shrinkage' must be more than 0")
+
+   if ((shrinkage < 0.001) && ((shrinkage > 1)))
+        stop("'shrinkage' must be between [0.001,1]")
+
+
+   if(bag.fraction <= 0)
+        stop("'bag.fraction' must be more than 0")
+
+   if (bag.fraction > 1)
+        stop("'shrinkage' must be (0,1]")
+
+
+   # if trace=TRUE, print out running time
+   if(trace) {
+        cat("Start model training\n")
+        starttime <- Sys.time()
+    }
+
+   # model training
+   if ((!is.dframe(X_train)) && (!is.darray(X_train))) { # for small data, load the whole data into every core
+      # system parameters are transfered into foreach
+      foreach(i, 1:nExecutor, function(dGBM_modeli=splits(dl_GBM_model,i), best.iter=splits(dbest.iter,i), x=X_train,y=Y_train, 
+                  n.trees=n.trees, distribution=distribution, interaction.depth=interaction.depth, n.minobsinnode=n.minobsinnode, 
+                  shrinkage=shrinkage, bag.fraction=bag.fraction) {
+      library(gbm)
+
+      if (distribution=="multinomial") {
+         y <- unlist(y) #  convert it back to "factor" for multinomial distribution
```

Comment:
I am puzzled why "y" is assumed to be a list, and how unlist() converts it to factor in a general case.